src/services/fetch/web_sys.rs

```diff
@@ -0,0 +1,605 @@
+            let dom_exception = DomException::from(err);
+            match dom_exception.code() {
+                DomException::ABORT_ERR => FetchError::Canceled,
+                DomException::NETWORK_ERR => FetchError::NetworkFailure,
+                _ => FetchError::FetchFailed,
+            }
+        })?;
+
+        if *self.active.borrow() {
+            Ok(WebResponse::from(response))
+        } else {
+            Err(FetchError::Canceled)
+        }
+    }
+
+    async fn get_data(&self, response: &WebResponse) -> Result<DATA, FetchError> {
+        let data_promise = if self.binary {
+            response.array_buffer()
+        } else {
+            response.text()
+        }
+        .map_err(|_| FetchError::InvalidResponse)?;
+
+        let data_result = JsFuture::from(data_promise).await;
+        if *self.active.borrow() {
+            data_result
+                .map_err(|_| FetchError::InvalidResponse)
+                .and_then(DATA::from_js)
+        } else {
+            Err(FetchError::Canceled)
+        }
+    }
+}
+
+fn build_request(parts: Parts, body: &JsValue) -> Result<WebRequest, Error> {
+    // Map headers into a Js `Header` type.
+    let header_list = parts
+        .headers
+        .iter()
+        .map(|(k, v)| {
+            Ok(Array::from_iter(&[
+                JsValue::from_str(k.as_str()),
+                JsValue::from_str(
+                    v.to_str()
+                        .map_err(|_| anyhow!("Unparsable request header"))?,
+                ),
+            ]))
+        })
+        .collect::<Result<Array, Error>>()?;
+
+    let header_map = Headers::new_with_str_sequence_sequence(&header_list)
+        .map_err(|_| anyhow!("couldn't build headers"))?;
+
+    // Formats URI.
+    let uri = parts.uri.to_string();
+    let method = parts.method.as_str();
+    let mut init = RequestInit::new();
+    init.method(method).body(Some(body)).headers(&header_map);
+    WebRequest::new_with_str_and_init(&uri, &init).map_err(|_| anyhow!("failed to build request"))
+}
+
+impl Task for FetchTask {
+    fn is_active(&self) -> bool {
+        if let Some(handle) = &self.0 {
+            *handle.active.borrow()
+        } else {
+            false
+        }
+    }
+
+    fn cancel(&mut self) {
+        // Fetch API doesn't support request cancelling in all browsers
+        // and we should use this workaround with a flag.
+        // In that case, request not canceled, but callback won't be called.
+        let handle = self
+            .0
+            .take()
+            .expect("tried to cancel request fetching twice");
+
+        *handle.active.borrow_mut() = false;
+        if let Some(abort_controller) = handle.abort_controller {
+            abort_controller.abort();
+        }
+    }
+}
+
+impl Drop for FetchTask {
+    fn drop(&mut self) {
+        if self.is_active() {
+            self.cancel();
+        }
+    }
+}
+
+thread_local! {
+    static GLOBAL: WindowOrWorker = WindowOrWorker::new();
```

Comment:
I replaced the `Reflect` implementation with this.
`Reflect` generates very unfavourable JS, it has a bit of overhead, because it requires transferring a string over the boundery. In comparison this implementation translates into pure JS returning the correct type.

Reply:
Great, I'm glad we aren't taking a performance hit due to Reflect for every fetch too. Looks good!